refactor(formatter): Refactor `try_print_flat` to `fits_on_line`

MichaReiser opened the pull request:
This PR brings the Printer closer to Prettier's printer by introducing a `fits_on_line` function that measures if printing this element won't exceed the line width.

This is somewhat similar to the existing `try_print_flat` implementation but differs in the sense that it also supports
measuring content that isn't printed in flat mode.

This functionality will be needed to implement alternatives because
the alternatives picks the first variant that fits on the line but
an alternative itself doesn't have to be printed in flat mode.


The PR doesn't change the semantics of `fits` yet so that only fits if the group's content fit as well as the remaining content in the document before the first new line character. 

This PR further introduces the concept of a `GroupId` and adds the ability to use `if_group_breaks` and `if_group_fits_on_line` with a specific group id. This is necessary because this PR aligns the `fill_elements` semantic with Prettier's by introducing an implicit group for each element. The problem of doing this is that the trailing comma insertion no longer worked because the element itself fits on the line (printed in flat mode) but the outer group of the array did break:

```
group([
  '[',
  indent([
    soft_line_break(),
    fill([
      [
        token('2222222'),
        if_group_breaks(',') // Doesn't get printed because the element fits
      ]
  ]),
  soft_line_break(),
])
```

MichaReiser commented on crates/rome_formatter/src/printer/printer.rs:
```diff
@@ -0,0 +1,1203 @@
+                        // print the group in "flat" mode, otherwise continue in expanded mode
+
+                        let flat_args = args.with_print_mode(PrintMode::Flat);
+                        // TODO replace `ElementCallQueue::default` with the real queue to also measure the rest of the document
```
I'll follow up on this TODO in #2579 Passing the rest of the document will make sure that e.g. the following snipped breaks if the `]` doesn't fit on the line:

```
[a, b, c]
```

MichaReiser commented on crates/rome_formatter/src/printer/mod.rs:
```diff
@@ -0,0 +1,1203 @@
+                        // print the group in "flat" mode, otherwise continue in expanded mode
+
+                        let flat_args = args.with_print_mode(PrintMode::Flat);
+                        // TODO replace `ElementCallQueue::default` with the real queue to also measure the rest of the document
```
This is part of #2579 Passing in the `queue` will make sure that `[a, b, c]` only fits if the `]` also fits on the line even if it's outside of the group.

MichaReiser replied:
`MeasureQueue` uses an offset into the queue to get the current position. It never actually dequeues.

MichaReiser commented on crates/rome_formatter/src/printer/mod.rs:
```diff
@@ -0,0 +1,1203 @@
+    }
+
+    pub fn with_incremented_indent(mut self) -> Self {
+        self.indent += 1;
```
Some comments have moved in this PR because `with_incremented_indent` only incremented the `indent` but didn't copy over the other fields like `hard_group`.

MichaReiser commented:
!bench_formatter

MichaReiser commented on crates/rome_formatter/src/printer/mod.rs:
```diff
@@ -0,0 +1,1177 @@
+    /// * The first and second content fit on a single line. It prints the content and separator in flat mode.
+    /// * The first content fits on a single line, but the second doesn't. It prints the content in flat and the separator in expanded mode.
+    /// * Neither the first nor the second content fit on the line. It brings the first content and the separator in expanded mode.
+    fn print_fill(
```
@NicholasLYang it should be straightforward for you to change the `fill` semantics so that it supports custom separators. Something I understand you need for JSX. This is prettier's implementation 

https://github.com/prettier/prettier/blob/bb130d2d6e6ea42629b17a57869f5e7e550df6fd/src/document/doc-printer.js#L415-L486

Make sure to also change the fits handling inside `fits_on_line`

MichaReiser commented:
!bench_formatter

leops commented on crates/rome_formatter/src/format_element.rs:
```diff
@@ -1181,26 +1252,103 @@ impl Deref for List {
+
+impl DebugGroupId {
+    pub fn new(debug_name: &'static str) -> Self {
+        static ID: AtomicU32 = AtomicU32::new(1);
```
I think it would be better to have the group ID allocation take place in the `Formatter`, in a long running process like the language server this static counter will eventually overflow and cause the `NonZeroU32` constructor below to panic (it sure would take a huge number of formatting runs for it to reach 2^32, but I'm also probably not the only one that never quits my editor and always just put my laptop to sleep, so it's not entirely unrealistic either ...)

MichaReiser replied:
That makes sense. I forgot that this will be across all files. I considered a `u32` to be sufficient for a single formatting run... So yes, let's move this to the formatter.

MichaReiser commented on crates/rome_formatter/src/format_element.rs:
```diff
@@ -1689,19 +1839,32 @@ mod tests {
-    }
 }
+
+#[cfg(target_pointer_width = "64")]
```
The macro is only defined in that configuration but I guess I can remove that restrictions

MichaReiser replied:
The checks don't seem to work for wasm. So I'll leave it as-is for now.

MichaReiser commented on crates/rome_formatter/src/printer/mod.rs:
```diff
@@ -0,0 +1,1177 @@
+    line_suffixes: Vec<PrintElementCall<'a>>,
+    verbatim_markers: Vec<TextRange>,
+    /// Tracks the mode in which groups with ids are printed.
+    group_modes: HashMap<GroupId, PrintMode>,
```
You mean by iterating over the groups instead? The challenge is that there's currently no "exit" event when the printer's done printing a group except if we use `print_all` but that makes the printer recursive which we want to avoid when possible. 

The only other alternative I see is to introduce a new IR element for this which is something I want to defer for now but would play into the *flat element buffer* proposal

leops replied:
I purely meant that using a `Vec` to store the group modes by using the GroupId as an index has constant-time lookups (and straightforward amortized insertion), and would be more efficient than using a `HashMap` (but it requires actually having small, densely allocated numerical keys or the memory usage will blow up)

MichaReiser commented:
Addressed the review feedback:
* The `PrinterState` now uses a `Vec<Option<GroupId>>` 
* Moved the creation of `GroupId`s to the `formatter so that the ids are scoped to a single document
* Rewrote `fill` loop to use a `for`
